[Do not merge yet] Typescript definitions

FrogTheFrog opened the pull request:
Here are some typescript definitions. Could you please check `TODO`s I have written? It contains some questions, bugs and other stuff I found curious.

jonschlinkert commented on index.d.ts:
```diff
@@ -0,0 +1,745 @@
+     */
+    //TODO: Which one is default or legacy?
+    value?: string;
+
```
Default is `value`. I thought I replaced all occurrences of `val`, but I just noticed at least one place in a code comment where `val` is used.

jonschlinkert commented on index.d.ts:
```diff
@@ -0,0 +1,745 @@
+     */
+    //TODO: Is this a legacy? Should it be removed?
+    unshiftNode?: (node: NodeLike<T>) => number;
+
```
snapdragon-utils has an `.unshiftNode` method, but nodes themselves (instances of `snapdragon-node`, at least newer versions) have an `.unshift` method. When the method is actually on an instance of node, we drop the `Node` part of the method name, since it's already implicit. 

Same naming conventions for the other methods.

FrogTheFrog replied:
`utils.addOpen` has a `node.unshiftNode` fallback method in it. That's why I thought that it was a legacy method. That's what got me confused, could it be a bug?

jonschlinkert commented on index.d.ts:
```diff
@@ -0,0 +1,745 @@
+     * [Optional] Getter that returns last child node.
+     */
+    //TODO: Can be undefined and null. Why not stick to undefined?
+    readonly last: NodeLike<T> | undefined | null;
```
> Why not stick to undefined?

In my experience, `undefined` generally means something is literally "not defined", whereas `null` is a value that was intentionally set. This distinction allows implementors to more easily know if something was intentional or not. 

I'm not opposed to simplifying that to `undefined`, but perhaps in the next major.

FrogTheFrog replied:
Oh, I totally understand that, however this (in snapdragon-node) really bugs me out:
`return this.nodes ? this.nodes[0] : null;`

Let's say that `this.node` is empty array. Since it is not undefined we will have `this.nodes[0]` returned which is undefined.

Correct solution would be:
`return this.nodes ? (this.nodes.length > 0 ? this.nodes[0] : null) : null;`
or, if we don't care about array containing actual undefined element:
`return this.nodes ? this.nodes[0] || null : null;`

jonschlinkert commented on index.d.ts:
```diff
@@ -0,0 +1,745 @@
+ */
+
+//TODO(self): Just return boolean?
+export function isNode<T>(node: NodeLike<T>): node is NodeLike<T>;
```
> Just return boolean?

correct

FrogTheFrog commented:
@jonschlinkert Ping pong. Also, take your time - lack of typescript definitions is just a minor inconvenience for me.

FrogTheFrog commented:
@jonschlinkert Hey, could you finish checking "TODOs" when you have time?

FrogTheFrog commented:
Ok, so what it did so far:

1. Rewrote everything in typescript while fixing stuff.
2. Compiled everything to javascript and changed tests.

I'm keeping typescript in this PR until it is ready to be merged (just in case something needs fixing). Used typescript in strict mode which went "nuclear" on me. For that reason I added a lot of type assertions to satisfy typescript, given that `utils` is meant to be a shim.

There were some cases where `if (!node) return null;` and similar code was used, I replaced them with type assertions, because it is not intended use of a function. I fixed test to reflect those changes.

@jonschlinkert there is one last simple `TODO` left, could you take a look at it?

jonschlinkert commented:
> there is one last simple TODO left, could you take a look at it?

excellent! I just got back from vacation. I'll take a look asap!
